src/createStore.js

```diff
@@ -163,38 +172,31 @@ export default function createStore(reducer, preloadedState, enhancer) {
    * return something else (for example, a Promise you can await).
    */
   function dispatch(action) {
-    if (!isPlainObject(action)) {
-      throw new Error(
-        'Actions must be plain objects. ' +
-        'Use custom middleware for async actions.'
-      )
-    }
-
-    if (typeof action.type === 'undefined') {
-      throw new Error(
-        'Actions may not have an undefined "type" property. ' +
-        'Have you misspelled a constant?'
-      )
-    }
-
-    if (isDispatching) {
-      throw new Error('Reducers may not dispatch actions.')
-    }
+    validateAction(action)
+    currentState = store.final.reducer(currentState, action)
+    store.final.onChange()
```

Comment:
@tappleby : that's really related to the intent of the core Redux subscription API itself.  There's been _numerous_ requests since the beginning to add either the action or the current state as an argument to subscription callbacks.  Issue #580 has Dan's reasoning and a roundup of almost all the other times this question was asked.

That said, I'm vaguely starting to wonder if we at least ought to re-discuss the idea for a notional 4.0 release.  I'd have to go back and review all the prior discussions, but I'm not immediately seeing actual harm that might happen if we implemented the idea.